jctools-core/src/main/java/org/jctools/queues/atomic/MpmcAtomicArrayQueue.java

```diff
@@ -72,17 +97,56 @@ protected final boolean casConsumerIndex(long expect, long newValue) {
     }
 }
 
+/**
+ * NOTE: This class was automatically generated by org.jctools.queues.atomic.JavaParsingAtomicArrayQueueGenerator
+ * which can found in the jctools-build module. The original source file is MpmcArrayQueue.java.
+ */
 abstract class MpmcAtomicArrayQueueL3Pad<E> extends MpmcAtomicArrayQueueConsumerIndexField<E> {
+
     long p01, p02, p03, p04, p05, p06, p07;
+
     long p10, p11, p12, p13, p14, p15, p16, p17;
+
     public MpmcAtomicArrayQueueL3Pad(int capacity) {
         super(capacity);
     }
 }
 
+/**
+ * NOTE: This class was automatically generated by org.jctools.queues.atomic.JavaParsingAtomicArrayQueueGenerator
+ * which can found in the jctools-build module. The original source file is MpmcArrayQueue.java.
+ 
+ * A Multi-Producer-Multi-Consumer queue based on a {@link ConcurrentCircularArrayQueue}. This implies that
+ * any and all threads may call the offer/poll/peek methods and correctness is maintained. <br>
+ * This implementation follows patterns documented on the package level for False Sharing protection.<br>
+ * The algorithm for offer/poll is an adaptation of the one put forward by D. Vyukov (See <a
+ * href="http://www.1024cores.net/home/lock-free-algorithms/queues/bounded-mpmc-queue">here</a>). The original
+ * algorithm uses an array of structs which should offer nice locality properties but is sadly not possible in
+ * Java (waiting on Value Types or similar). The alternative explored here utilizes 2 arrays, one for each
+ * field of the struct. There is a further alternative in the experimental project which uses iteration phase
+ * markers to achieve the same algo and is closer structurally to the original, but sadly does not perform as
+ * well as this implementation.<br>
+ *
+ * Tradeoffs to keep in mind:
+ * <ol>
+ * <li>Padding for false sharing: counter fields and queue fields are all padded as well as either side of
+ * both arrays. We are trading memory to avoid false sharing(active and passive).
+ * <li>2 arrays instead of one: The algorithm requires an extra array of longs matching the size of the
+ * elements array. This is doubling/tripling the memory allocated for the buffer.
+ * <li>Power of 2 capacity: Actual elements buffer (and sequence buffer) is the closest power of 2 larger or
+ * equal to the requested capacity.
+ * </ol>
+ *
+ * @param <E>
+ *            type of the element stored in the {@link java.util.Queue}
+ */
 public class MpmcAtomicArrayQueue<E> extends MpmcAtomicArrayQueueL3Pad<E> {
-    
-    public MpmcAtomicArrayQueue(int capacity) {
+
+    static final int RECOMENDED_POLL_BATCH = CPUs * 4;
+
+    public static final int RECOMENDED_OFFER_BATCH = CPUs * 4;
```

Comment:
These two static fields are used by the messaging passing method bodies. I haven't done anything to remove these. I don't know if I should as they are accessible by users?